### PR TITLE
[Merged by Bors] - feat: port Order.Filter.Germ

### PR DESCRIPTION
**Pull request by Ruben-VandeVelde**
---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by nomeata** [comment]
I tried to help out, but gut stuck with `to_additive` not translating `instPowGerm` in `instance monoid`, and writing `instance addMonoid` by hand also failed because of coercions getting strangly in the way.

File: Mathlib/Order/Filter/Germ.lean [hunk 1]
```diff
@@ -0,0 +1,728 @@
...
+
+@[elab_as_elim]
+theorem inductionOn₃ (f : Germ l β) (g : Germ l γ) (h : Germ l δ)
+    {p : Germ l β → Germ l γ → Germ l δ → Prop}
+    (H : ∀ (f : α → β) (g : α → γ) (h : α → δ), p f g h) : p f g h :=
+  Quotient.inductionOn₃' f g h H
+#align filter.germ.induction_on₃ Filter.Germ.inductionOn₃
+
+/-- Given a map `F : (α → β) → (γ → δ)` that sends functions eventually equal at `l` to functions
+eventually equal at `lc`, returns a map from `germ l β` to `germ lc δ`. -/
```
Comment (dupuisf) [inline Mathlib/Order/Filter/Germ.lean:154]:
```suggestion
/-- Given a map `F : (α → β) → (γ → δ)` that sends functions eventually equal at `l` to functions
eventually equal at `lc`, returns a map from `Germ l β` to `Germ lc δ`. -/
```

File: Mathlib/Order/Filter/Germ.lean [hunk 2]
```diff
@@ -0,0 +1,728 @@
...
+#align filter.germ.map₂ Filter.Germ.map₂
+
+@[simp]
+theorem map₂_coe (op : β → γ → δ) (f : α → β) (g : α → γ) :
+    map₂ op (f : Germ l β) g = fun x => op (f x) (g x) :=
+  rfl
+#align filter.germ.map₂_coe Filter.Germ.map₂_coe
+
+/-- A germ at `l` of maps from `α` to `β` tends to `lb : filter β` if it is represented by a map
+which tends to `lb` along `l`. -/
```
Comment (dupuisf) [inline Mathlib/Order/Filter/Germ.lean:215]:
```suggestion
/-- A germ at `l` of maps from `α` to `β` tends to `lb : Filter β` if it is represented by a map
which tends to `lb` along `l`. -/
```

File: Mathlib/Order/Filter/Germ.lean [hunk 3]
```diff
@@ -0,0 +1,728 @@
...
+#align filter.germ.comp_tendsto' Filter.Germ.compTendsto'
+
+@[simp]
+theorem coe_compTendsto' (f : α → β) {lc : Filter γ} {g : Germ lc α} (hg : g.Tendsto l) :
+    (f : Germ l β).compTendsto' g hg = g.map f :=
+  rfl
+#align filter.germ.coe_comp_tendsto' Filter.Germ.coe_compTendsto'
+
+/-- Given a germ `f : germ l β` and a function `g : γ → α`, where `l : filter α`, if `g` tends
+to `l` along `lc : filter γ`, then the composition `f ∘ g` is well-defined as a germ at `lc`. -/
```
Comment (dupuisf) [inline Mathlib/Order/Filter/Germ.lean:242]:
```suggestion
/-- Given a germ `f : Germ l β` and a function `g : γ → α`, where `l : Filter α`, if `g` tends
to `l` along `lc : Filter γ`, then the composition `f ∘ g` is well-defined as a germ at `lc`. -/
```

File: Mathlib/Order/Filter/Germ.lean [hunk 4]
```diff
@@ -0,0 +1,728 @@
...
+@[simp, norm_cast]
+theorem coe_tendsto {f : α → β} {lb : Filter β} : (f : Germ l β).Tendsto lb ↔ Tendsto f l lb :=
+  Iff.rfl
+#align filter.germ.coe_tendsto Filter.Germ.coe_tendsto
+
+alias coe_tendsto ↔ _ _root_.Filter.Tendsto.germ_tendsto
+#align filter.tendsto.germ_tendsto Filter.Tendsto.germ_tendsto
+
+/-- Given two germs `f : germ l β`, and `g : germ lc α`, where `l : filter α`, if `g` tends to `l`,
+then the composition `f ∘ g` is well-defined as a germ at `lc`. -/
```
Comment (dupuisf) [inline Mathlib/Order/Filter/Germ.lean:229]:
```suggestion
/-- Given two germs `f : Germ l β`, and `g : Germ lc α`, where `l : Filter α`, if `g` tends to `l`,
then the composition `f ∘ g` is well-defined as a germ at `lc`. -/
```

File: Mathlib/Order/Filter/Germ.lean [hunk 5]
```diff
@@ -0,0 +1,728 @@
...
+#align filter.germ.const_comp_tendsto Filter.Germ.const_compTendsto
+
+@[simp]
+theorem const_compTendsto' {l : Filter α} (b : β) {lc : Filter γ} {g : Germ lc α}
+    (hg : g.Tendsto l) : (const l b).compTendsto' g hg = const _ b :=
+  inductionOn g (fun _ _ => rfl) hg
+#align filter.germ.const_comp_tendsto' Filter.Germ.const_compTendsto'
+
+/-- Lift a predicate on `β` to `germ l β`. -/
+def LiftPred (p : β → Prop) (f : Germ l β) : Prop :=
```
Comment (dupuisf) [inline Mathlib/Order/Filter/Germ.lean:289]:
```suggestion
/-- Lift a predicate on `β` to `Germ l β`. -/
def LiftPred (p : β → Prop) (f : Germ l β) : Prop :=
```

File: Mathlib/Order/Filter/Germ.lean [hunk 6]
```diff
@@ -0,0 +1,728 @@
...
+  eventually_of_forall fun _y => hx
+#align filter.germ.lift_pred_const Filter.Germ.liftPred_const
+
+@[simp]
+theorem liftPred_const_iff [NeBot l] {p : β → Prop} {x : β} : LiftPred p (const l x) ↔ p x :=
+  @eventually_const _ _ _ (p x)
+#align filter.germ.lift_pred_const_iff Filter.Germ.liftPred_const_iff
+
+/-- Lift a relation `r : β → γ → Prop` to `germ l β → germ l γ → Prop`. -/
+def LiftRel (r : β → γ → Prop) (f : Germ l β) (g : Germ l γ) : Prop :=
```
Comment (dupuisf) [inline Mathlib/Order/Filter/Germ.lean:309]:
```suggestion
/-- Lift a relation `r : β → γ → Prop` to `Germ l β → Germ l γ → Prop`. -/
def LiftRel (r : β → γ → Prop) (f : Germ l β) (g : Germ l γ) : Prop :=
```

File: Mathlib/Order/Filter/Germ.lean [hunk 7]
```diff
@@ -0,0 +1,728 @@
...
+#align filter.germ.coe_sub Filter.Germ.coe_sub
+
+-- porting notes: Removed `norm_cast`.
+@[to_additive (attr := simp)]
+theorem const_div [Div M] (a b : M) : const l (a / b) = (const l a) / (const l b) :=
+  rfl
+#align filter.germ.const_div Filter.Germ.const_div
+#align filter.germ.const_sub Filter.Germ.const_sub
+
+-- Porting note: `to_addive` can't generate this.
```
Comment (jcommelin) [inline Mathlib/Order/Filter/Germ.lean:484]:
```suggestion
-- Porting note: `to_additive` can't generate this.
```

File: Mathlib/Order/Filter/Germ.lean [hunk 8]
```diff
@@ -239,76 +238,77 @@ theorem coe_compTendsto' (f : α → β) {lc : Filter γ} {g : Germ lc α} (hg :
...
 
 @[simp]
 theorem coe_compTendsto (f : α → β) {lc : Filter γ} {g : γ → α} (hg : Tendsto g lc l) :
-    (f : Germ l β).comp_tendsto g hg = f ∘ g :=
+    (f : Germ l β).compTendsto g hg = f ∘ g :=
   rfl
 #align filter.germ.coe_comp_tendsto Filter.Germ.coe_compTendsto
 
-@[simp]
+-- @[simp] -- Porting note: simp cannot prove this
```
Comment (jcommelin) [inline Mathlib/Order/Filter/Germ.lean:253]: If simp cannot prove this, please don't comment out the `@[simp]` attribute.
Instead, if the linter complains, add `nolint simpNF`.
```suggestion
@[simp, nolint simpNF] -- Porting note: simp cannot prove this
```

File: Mathlib/Order/Filter/Germ.lean [hunk 9]
```diff
@@ -239,76 +238,77 @@ theorem coe_compTendsto' (f : α → β) {lc : Filter γ} {g : Germ lc α} (hg :
...
+    f.compTendsto' _ hg.germ_tendsto = f.compTendsto g hg :=
   rfl
 #align filter.germ.comp_tendsto'_coe Filter.Germ.compTendsto'_coe
 
-@[simp, norm_cast]
-theorem const_inj [NeBot l] {a b : β} : (↑a : Germ l β) = ↑b ↔ a = b :=
-  coe_eq.trans <| const_eventually_eq
+-- Porting note: used `const` instead of lift
+@[simp] -- Porting note: removed norm_cast
+theorem const_inj [NeBot l] {a b : β} : const l a = const l b ↔ a = b :=
```
Comment (jcommelin) [inline Mathlib/Order/Filter/Germ.lean:261]: Why is this needed? Can't we make the coercion work?
I don't know how many proofs in downstream file make use of `norm_cast`... but they would all break.

Comment (Komyyy) [reply Mathlib/Order/Filter/Germ.lean:261]: I think this is an alternative of `has_lift_t`, but as you say, we had better protect proofs using `norm_cast`. I'll replace `const` to a coercion and re-add `norm_cast`.

File: Mathlib/Order/Filter/Germ.lean [hunk 10]
```diff
@@ -239,27 +243,28 @@ theorem coe_compTendsto' (f : α → β) {lc : Filter γ} {g : Germ lc α} (hg :
...
 
-@[simp]
+@[simp, nolint simpNF] -- Porting note: simp cannot prove this
 theorem compTendsto'_coe (f : Germ l β) {lc : Filter γ} {g : γ → α} (hg : Tendsto g lc l) :
-    f.compTendsto' _ hg.germ_tendsto = f.comp_tendsto g hg :=
+    f.compTendsto' _ hg.germ_tendsto = f.compTendsto g hg :=
   rfl
 #align filter.germ.comp_tendsto'_coe Filter.Germ.compTendsto'_coe
 
+-- Porting note: used `const` instead of lift
```
Comment (jcommelin) [inline Mathlib/Order/Filter/Germ.lean:264]:
```suggestion

```

**Review by jcommelin — APPROVED** [review]
Thanks!

bors d+

**Comment by jcommelin** [comment]
bors d=Komyyy

**Pull request closed without merging**